components/sync15/src/key_bundle.rs

```diff
@@ -75,53 +77,35 @@ impl KeyBundle {
         [base64::encode(&self.enc_key), base64::encode(&self.mac_key)]
     }
 
-    /// Returns the 32 byte digest by value since it's small enough to be passed
-    /// around cheaply, and easily convertable into a slice or vec if you want.
-    fn hmac(&self, ciphertext: &[u8]) -> Result<[u8; 32]> {
-        let mut out = [0u8; 32];
-        let key = PKey::hmac(self.hmac_key())?;
-        let mut signer = Signer::new(MessageDigest::sha256(), &key)?;
-        signer.update(ciphertext)?;
-        let size = signer.sign(&mut out)?;
-        // This isn't an Err since it really should not be possible.
-        assert!(
-            size == 32,
-            "Somehow the 256 bits from sha256 do not add up into 32 bytes..."
-        );
-        Ok(out)
+    fn hmac(&self, ciphertext: &[u8]) -> Result<Signature> {
+        let key = SigningKey::new(&digest::SHA256, self.hmac_key());
+        Ok(hmac::sign(&key, ciphertext)?)
     }
 
     /// Important! Don't compare against this directly! use `verify_hmac` or `verify_hmac_string`!
     pub fn hmac_string(&self, ciphertext: &[u8]) -> Result<String> {
         Ok(base16::encode_lower(&self.hmac(ciphertext)?))
     }
 
-    pub fn verify_hmac(&self, expected_hmac: &[u8], ciphertext_base64: &str) -> Result<bool> {
-        let computed_hmac = self.hmac(ciphertext_base64.as_bytes())?;
-        // I suspect this is unnecessary for our case, but the rust-openssl docs
-        // want us to use this over == to avoid sidechannels, and who am I to argue?
-        Ok(openssl::memcmp::eq(&expected_hmac, &computed_hmac))
+    pub fn verify_hmac(&self, expected_hmac: &[u8], ciphertext_base64: &str) -> Result<()> {
+        let key = VerificationKey::new(&digest::SHA256, self.hmac_key());
+        Ok(hmac::verify(
+            &key,
+            ciphertext_base64.as_bytes(),
+            expected_hmac,
+        )?)
     }
 
-    pub fn verify_hmac_string(&self, expected_hmac: &str, ciphertext_base64: &str) -> Result<bool> {
-        let computed_hmac = self.hmac(ciphertext_base64.as_bytes())?;
-        // Note: openssl::memcmp::eq panics if the sizes aren't the same. Desktop returns that it
-        // was a verification failure, so we will too.
-        if expected_hmac.len() != 64 {
-            log::warn!("Garbage HMAC verification string: Wrong length");
-            return Ok(false);
-        }
+    pub fn verify_hmac_string(&self, expected_hmac: &str, ciphertext_base64: &str) -> Result<()> {
         // Decode the expected_hmac into bytes to avoid issues if a client happens to encode
         // this as uppercase. This shouldn't happen in practice, but doing it this way is more
         // robust and avoids an allocation.
         let mut decoded_hmac = [0u8; 32];
-
         if base16::decode_slice(expected_hmac, &mut decoded_hmac).is_err() {
             log::warn!("Garbage HMAC verification string: contained non base16 characters");
-            return Ok(false);
```

Comment:
I am very, very pleased to see this throwing an error rather than returning `OK(false)`, excellent cleanup!